tensorflow/core/common_runtime/dml/dml_execution_context.cc

```diff
@@ -372,4 +421,176 @@ void DmlExecutionContextImpl::CloseCommandListAndExecute() {
+    absl::Span<const uint8_t> value) {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+
+  absl::InlinedVector<uint8_t, 16> value_copy(value.size());
+  std::copy(value.begin(), value.end(), value_copy.begin());
+
+  shared_state_->WriteBatch().emplace_back(
+      [=, value_copy = std::move(value_copy)]() {
+        shared_state_->impl->FillBufferWithPattern(
+            dst, dst_offset, dst_size_in_bytes, value_copy);
+      });
+
+  return shared_state_->next_flush_event;
+}
+
+DmlGpuEvent DmlExecutionContext::InitializeOperator(
+    IDMLOperatorInitializer* initializer,
+    Microsoft::WRL::ComPtr<IDMLBindingTable>&& binding_table,
+    ID3D12DescriptorHeap* descriptor_heap) {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+
+  shared_state_->WriteBatch().emplace_back(
+      [=, binding_table = std::move(binding_table)]() {
+        shared_state_->impl->InitializeOperator(
+            initializer, binding_table.Get(), descriptor_heap);
+      });
+
+  shared_state_->new_function_enqueued.notify_all();
+
+  return shared_state_->next_flush_event;
+}
+
+DmlGpuEvent DmlExecutionContext::ExecuteOperator(
+    IDMLCompiledOperator* op,
+    Microsoft::WRL::ComPtr<IDMLBindingTable>&& binding_table,
+    ID3D12DescriptorHeap* descriptor_heap) {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+
+  shared_state_->WriteBatch().emplace_back(
+      [=, binding_table = std::move(binding_table)]() {
+        shared_state_->impl->ExecuteOperator(op, binding_table.Get(),
+                                             descriptor_heap);
+      });
+
+  shared_state_->new_function_enqueued.notify_all();
+
+  return shared_state_->next_flush_event;
+}
+
+DmlGpuEvent DmlExecutionContext::ResourceBarrier(
+    absl::Span<const D3D12_RESOURCE_BARRIER> barriers) {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+
+  // The caller may not keep the barriers referenced by the span alive for
+  // longer than this function call, so make a copy and transfer ownership to
+  // the lambda.
+  absl::InlinedVector<D3D12_RESOURCE_BARRIER, 4> barriers_copy;
+  shared_state_->WriteBatch().emplace_back(
+      [=, barriers = std::move(barriers_copy)]() {
+        shared_state_->impl->ResourceBarrier(barriers);
+      });
+
+  shared_state_->new_function_enqueued.notify_all();
+
+  return shared_state_->next_flush_event;
+}
+
+DmlGpuEvent DmlExecutionContext::UavBarrier() {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+
+  shared_state_->WriteBatch().emplace_back(
+      [=]() { shared_state_->impl->UavBarrier(); });
+
+  shared_state_->new_function_enqueued.notify_all();
+
+  return shared_state_->next_flush_event;
+}
+
+StatusOr<DmlGpuEvent> DmlExecutionContext::Flush() {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+
+  auto event = shared_state_->next_flush_event;
+  if (shared_state_->WriteBatch().empty()) {
+    --event.fence_value;
+  }
+
+  shared_state_->flush_requested = true;
```

Comment:
Won't this result in a flush that isn't well ordered with respect to items being added to the batch? That is, the actual flush could occur at any point arbitrarily far in the future (after some other stuff gets added to the batch, for example). Are there any scenarios where this could cause a problem?

Reply:
It just never ends up mattering. Flush can't be relied on for CPU/GPU synchronization unless you use the returned GPU event, which by definition means you only wait on stuff already in the batch, which obviously works just fine. For GPU/GPU synchronization you'd also need to use fences or barriers, and the implicit assumption is that kernels execute ops in the correct order (or the order just doesn't matter).

The only times we call Flush externally are for readbacks (just to ensure the GPU work commences) or device sync.